docs/it/refactoring.md

```diff
@@ -7,3 +7,132 @@
+Da un codice scritto bene, si possono ottenere migliori performance. Magari non direttamente ma, una buona attività di refactoring ci permetterà di avere molto più chiaro come il software interagisce, dandoci la possibilità di individuare eventuali colli di bottiglia, anche a livello architetturale.
+
+## Quando
+
+La scrittura del codice non è del tutto lineare, soprattutto se scritto a più mani. Ogni sviluppatore ha un suo stile, delle sue convinzioni, un determinato livello di preparazione. Questo porta spesso ad avere del codice, scritto a colpi di commit, funzionante ma difficile da comprendere, con scorciatoie evitabili, metodi prolissi o, peggio ancora, costanti che non dovrebbero esserlo.
+Sono diverse le occasioni in cui è possibile fare refactoring e, in alcune di queste, non solo è consigliato ma obbligatorio.
+Poniamo il caso che abbiamo una parte del nostro software con del codice "legacy", ovvero scritto più di tre mesi fà, che funziona egregiamente, fa il suo ma, ci rendiamo conto che le performance non sono quelle che ci attendiamo.
+Fatta una rapida analisi, notiamo che, magari, ci sono dei cicli ripetuti, o parti di codice che potrebbero essere evitate se solo avessimo posto dei semafori a monte.
+Con del buon refactoring potremmo rivedere il codice, eliminare delle dipendenze inutili o utilizzare dei costrutti del linguaggio più adatti, con il risultato di migliorare le performance.
+E se non sono le performance a preoccuparci, ma la poltiglia di codice scritto, magari di fretta e sotto pressione, che non rispetta principi base, come la responsabilità singola? Anche in questo caso, possiamo rivedere il codice, renderlo più leggibile e comprensibile così da ottenere un altro bel risultato: eliminare debito tecnico.
+Ma, come detto, le occasioni non mancano.
+Potremmo cogliere l'occasione di una Code Review o l'aggiunta di una nuova feature. Prima di iniziarne lo sviluppo, sicuramente è buona pratica leggere il codice in cui questa nuova funzionalità andrà a integrarsi e, se quest'ultimo risulta di difficile comprensione, è un'ottima occasione per effettuarne la rifattorizzazione. Questo ci permetterebbe di ottenere una maggiore conoscienza del codice.
+Un'altra, forse la migliore, dove è (quasi) obbligatorio, è la risoluzione di un bug in una parte di codice convulso, la cui sola lettura ci fa venire il mal di testa. Un'attività di refactoring potrebbe facilmente portare alla luce l'errore presente nel codice.
+L'attività di refactoring può essere paragonata a una pulizia profonda di casa, piuttosto che a una routine di sistemazione della propria cameretta, necessaria per mantenere il codice sicuro, di valore ed efficiente.
+
+## Quando è meglio evitare
+
+Non sempre è utile effettuare refactoring. Ci sono casi in cui il codice è già perfettamente ottimizzato anche se, a una vista inesperta, potrebbe risultare di difficile lettura.
```

Comment:
Più che spingere sulla dicotomia performance-capacità di lettura, direi che ci sono contesti in cui la lettura e la maintenance hanno un valore maggiore rispetto alle performance e che quindi pur avendo idea di come migliorare il codice in termini di performance, lasciamo stare perché la leggibilità ha un valore maggiore. Ricordo una mia PR che migliorava il tempo di build medio di Vite del 3% che è stata rifiutata perché la manutenzione di quella parte di codice sarebbe risultata più complessa.

Reply:
Aggiunto accenno alla mantenibilità ed al TDD